Add fetch job and update stage_ic to work with fetched ICs

DavidGrumm-NOAA opened the pull request:
# Description
Most jobs require the initial conditions to be available on local disk. The existing “stage_ic” task copies/stages these initial condition into the experiment's COM directory. This PR for the “fetch” task extends that functionality to copy from HPSS (on HPSS-accessible machines) into COM.

Resolves #2988

# Type of change
- [X] New feature (adds functionality)

# Change characteristics
- Is this a breaking change (a change in existing functionality)? NO
- Does this change require a documentation update? YES
- Does this change require an update to any of the following submodules? NO

# How has this been tested?
 - In process of being tested on Hera

# Checklist
- [ ] Any dependent changes have been merged and published
- [X] My code follows the style guidelines of this project
- [ ] I have performed a self-review of my own code
- [X] I have commented my code, particularly in hard-to-understand areas
- [] I have documented my code, including function, input, and output descriptions
- [X] My changes generate no new warnings
- [ ] New and existing tests pass with my changes
- [ ] This change is covered by an existing CI test or a new one has been added
- [X] Any new scripts have been added to the .github/CODEOWNERS file with owners
- [ ] I have made corresponding changes to the system documentation if necessary

DavidGrumm-NOAA committed:
Stage_ic updates: GH2988

DavidGrumm-NOAA commented:
I am in the process of testing.

DavidGrumm-NOAA commented:
To test my code, I ran create_experiment with the short yaml C48_ATM.yaml, (which created /scratch1/NCEPDEV/global/David.Grumm/G_WF_2988/testroot_1/EXPDIR and COMROOT) by :

HPC_ACCOUNT="fv3-cpu" MY_TESTROOT="/scratch1/NCEPDEV/global/David.Grumm/G_WF_2988/testroot_1" RUNTESTS=${MY_TESTROOT} pslot="1306a_2988" ./create_experiment.py --yaml ../ci/cases/pr/C48_ATM.yaml

… which completed without error or warning messages.

From within that EXPDIR I ran rocotorun:
rocotorun -w ./1306a_2988/1306a_2988.xml -d ./1306a_2988/1306a_2988.db

… which completed without error or warning messages. There was also no output to stdout, which I did not expect as I had placed a few diagnostic prints in my code. I verified that I am my current branch.

Runniing rocotostat gives me:

   CYCLE                    TASK                       JOBID               STATE         EXIT STATUS     TRIES      DURATION

========================================================================
202103231200 gfs_stage_ic druby://10.184.8.62:37937 SUBMITTING - 0 0.0
202103231200 gfs_fcst_seg0 - - - - -
202103231200 gfs_atmos_prod_f000 - - - - -
etc.
… and this appears unchanged (at least for the 4 hours since I ran rocotorun)

I have 2 questions:

   -  Am I not running my code, as the diagnostic prints do not appear ? ( I was able to run similarly modified versions of exglobal_fetch.py and fetch.py earlier when I had run fetch.sh, but I was running that script from the same directory. It simply may be a matter of manually resetting my PATH)
   
  - Rocotostat indicates that a job has been submitted (presumably with another version of the code)
    Shouldn’t I expect it to progress ?

DavidHuber-NOAA commented:
Rocoto is not a fully automated system.  For each invocation of `rocotorun`, it checks the status of running jobs and updates its database accordingly, checks to see if any new jobs have met their prerequisites, and submits jobs to the queue that are ready to run.  `rocotostat` simply reads the database and reports its contents.  Thus, `rocotorun` must be run every few minutes to continuously submit jobs.  Conveniently, in your `EXPDIR`, there is a file with extension `.crontab`.  Copy the contents of this file to your crontab (which you can edit via the `crontab -e` command).  Now your experiment should run continuously.  For more details, [read the docs](https://global-workflow.readthedocs.io/en/latest/start.html).

DavidHuber-NOAA commented on ush/python/pygfs/task/fetch.py:
```diff
@@ -0,0 +1,98 @@
+#!/usr/bin/env python3
+
+import os
+from logging import getLogger
+from typing import Any, Dict, List
+
+from wxflow import (AttrDict, FileHandler, Hsi, Task,
+                    logit, parse_j2yaml) 
+from wxflow import htar as Htar
+import tarfile  
```
```suggestion
import os
from logging import getLogger
from typing import Any, Dict, List

from wxflow import (AttrDict, FileHandler, Hsi, Task,
                    logit, parse_j2yaml)
from wxflow import htar as Htar
import tarfile 
```

DavidGrumm-NOAA replied:
Done

DavidHuber-NOAA commented on ush/python/pygfs/task/fetch.py:
```diff
@@ -0,0 +1,98 @@
+        None
+        """
+        super().__init__(config)
+        
```
```suggestion

```

DavidGrumm-NOAA replied:
Done

DavidHuber-NOAA commented on ush/python/pygfs/task/fetch.py:
```diff
@@ -0,0 +1,98 @@
+        fetch_parm = os.path.join(fetch_dict.PARMgfs, "fetch")
+
+        parsed_fetch = parse_j2yaml(os.path.join(fetch_parm, fetch_yaml),
+	                            fetch_dict)
```
```suggestion
        parsed_fetch = parse_j2yaml(os.path.join(fetch_parm, fetch_yaml),
                                                 fetch_dict)
```

DavidHuber-NOAA commented on ush/python/pygfs/task/fetch.py:
```diff
@@ -0,0 +1,98 @@
+
+        parsed_fetch = parse_j2yaml(os.path.join(fetch_parm, fetch_yaml),
+	                            fetch_dict)
+        
```
```suggestion

```

DavidGrumm-NOAA replied:
Done

DavidHuber-NOAA commented on ush/python/pygfs/task/fetch.py:
```diff
@@ -0,0 +1,98 @@
+
+    @logit(logger)
+    def execute_pull_data(self, atardir_set: Dict[str, Any]) -> None:
+        """Pull data from HPSS based on a yaml dictionary and store at the 
```
```suggestion
        """Pull data from HPSS based on a yaml dictionary and store at the
```

DavidGrumm-NOAA replied:
Done

DavidHuber-NOAA commented on ush/python/pygfs/task/fetch.py:
```diff
@@ -0,0 +1,98 @@
+
+        Parameters
+        ----------
+        atardir_set: Dict[str, Any],         
```
```suggestion
        atardir_set: Dict[str, Any],
```

DavidHuber-NOAA commented on ush/python/pygfs/task/fetch.py:
```diff
@@ -0,0 +1,98 @@
+            None
+        """
+        if len(f_names) <= 0:     # Abort if no files
+            raise FileNotFoundError("FATAL ERROR: The tar ball has no files") # DG - add name
```
```suggestion
            raise FileNotFoundError("FATAL ERROR: The tar ball has no files")  # DG - add name
```

DavidHuber-NOAA commented on ush/python/pygfs/task/fetch.py:
```diff
@@ -0,0 +1,98 @@
+        """
+        if len(f_names) <= 0:     # Abort if no files
+            raise FileNotFoundError("FATAL ERROR: The tar ball has no files") # DG - add name
+        
```
```suggestion

```

DavidGrumm-NOAA replied:
Done

DavidHuber-NOAA commented on ush/python/pygfs/task/fetch.py:
```diff
@@ -0,0 +1,98 @@
+        f_names = atardir_set.untar.contents
+        on_hpss = atardir_set.untar.on_hpss
+        dest = atardir_set.untar.destination
+    
```
```suggestion
```

DavidGrumm-NOAA replied:
Done

DavidHuber-NOAA commented on ush/python/pygfs/task/fetch.py:
```diff
@@ -0,0 +1,98 @@
+        dest = atardir_set.untar.destination
+    
+        # Select action whether no_hpss is True or not, and pull these data from
+        #    tape or locally and place where it needs to go        
+        # DG - these need testing
+        if on_hpss is True: # htar all files in fnames
```
```suggestion
        #    tape or locally and place where it needs to go
        # DG - these need testing
        if on_hpss is True:  # htar all files in fnames
```

DavidGrumm-NOAA replied:
Done

DavidHuber-NOAA commented on ush/python/pygfs/task/fetch.py:
```diff
@@ -0,0 +1,98 @@
+        # DG - these need testing
+        if on_hpss is True: # htar all files in fnames
+            htar_obj = Htar.Htar()
+            htar_obj.cvf(dest, f_names) 
+
+        else: # tar all files in fnames
```
```suggestion
            htar_obj.cvf(dest, f_names)

        else:  # tar all files in fnames
```

DavidGrumm-NOAA replied:
Done

DavidHuber-NOAA commented on ush/python/pygfs/task/fetch.py:
```diff
@@ -0,0 +1,98 @@
+            with tarfile.open(dest, "w") as tar:
+                for filename in f_names:
+                    tar.add(filename)
+
+
+    # Other helper methods...
+    
```
```suggestion
```

DavidGrumm-NOAA replied:
hmm .. I don't see a difference here,

DavidHuber-NOAA replied:
It's deleting the last 4 lines of the Python module since there weren't any other helper methods.

DavidGrumm-NOAA replied:
Sure, of course.

DavidHuber-NOAA reviewed (commented):
Just some whitespace cleanup (most of it mine).

DavidGrumm-NOAA commented:
I updated the crontab.

DavidGrumm-NOAA committed:
Removed white space

DavidGrumm-NOAA commented:
Removed extraneous white space from fetch.py and recommitted; still testing.

DavidGrumm-NOAA committed:
remove more whitespace

DavidGrumm-NOAA committed:
Incorporate fetch into GW

DavidGrumm-NOAA committed:
Merge branch 'develop' into stage_ic_2988

DavidGrumm-NOAA committed:
Fix whitespace

DavidGrumm-NOAA committed:
Reconcile divergent branches. Merge branch 'stage_ic_2988' of github.com:DavidGrumm-NOAA/Global_Workflow_2988 into stage_ic_2988

DavidGrumm-NOAA committed:
Move the fetch options to the run_options dict

DavidGrumm-NOAA commented:
I moved the fetch options to be in the run_options dict.

DavidHuber-NOAA commented on ci/cases/pr/ATM_cold.yaml:
I think this should be in `parm/fetch/ATM_cold.yaml.j2`

EDIT: added `.j2` as this is `Jinja`-templated.

DavidGrumm-NOAA replied:
Per the comment below, removing parm/fetch/ATM_cold.yaml.j2

DavidHuber-NOAA replied:
Sorry for the confusion, @DavidGrumm-NOAA.  I was suggesting that this file should be moved to `parm/fetch/ATM_cold.yaml.j2`.  This is the template for the fetch job, not for a CI test.

DavidGrumm-NOAA replied:
Done

DavidHuber-NOAA replied:
Thanks for making a copy to parm/fetch/ATM_cold.yaml.j2.  This file can be deleted now with `git rm ci/cases/pr/ATM_cold.yaml && git commit && git push`.

DavidHuber-NOAA commented on parm/config/gefs/config.fetch:
```diff
@@ -0,0 +1,38 @@
+export ICSDIR="@ICSDIR@" # User provided ICSDIR; blank if not provided
+export BASE_IC="@BASE_IC@" # Platform home for staged ICs
+
+export FETCH_YAML_TMPL="${PARMgfs}/stage/master_gefs.yaml.j2"
```
This should probably be
```suggestion
export FETCH_YAML_TMPL="${PARMgfs}/fetch/ATM_Cold.yaml.j2"
```

DavidHuber-NOAA replied:
N/M.  We don't have a fetch YAML for GEFS.  I think this file should be removed instead.

DavidGrumm-NOAA replied:
deleting that line from  parm/config/gefs/config.fetch

DavidHuber-NOAA commented on parm/config/gefs/config.fetch:
```diff
@@ -0,0 +1,38 @@
+export BASE_IC="@BASE_IC@" # Platform home for staged ICs
+
+export FETCH_YAML_TMPL="${PARMgfs}/stage/master_gefs.yaml.j2"
+export fetch_yaml="stage_atm_cold.yaml"
+
+# Set ICSDIR (if not defined)
+
+if [[ -z "${ICSDIR}" ]] ; then
+
+  ic_ver="20240610"
+
+  if (( NMEM_ENS > 0 )) ; then
+    ensic="${CASE_ENS}"   
+  fi
+
+  if [[ "${DO_OCN:-NO}" == "YES" ]] ; then
+    ocnic="mx${OCNRES}"
+  fi
+
+  export ICSDIR="${BASE_IC}/${CASE}${ensic:-}${ocnic:-}/${ic_ver}"
+
+fi
+
+# Use of perturbations files for ensembles
+export USE_OCN_ENS_PERTURB_FILES="NO"
+export USE_ATM_ENS_PERTURB_FILES="NO"
```
I don't think you need this.  These are all `stage_ic` variables.

```suggestion
```

DavidGrumm-NOAA replied:
Ok, deleted those files

DavidHuber-NOAA commented on parm/config/gfs/config.base:
```diff
@@ -474,6 +474,14 @@ export DO_VRFY_OCEANDA="@DO_VRFY_OCEANDA@"  # Run SOCA Ocean and Seaice DA verif
 [[ "${FHMAX_FITS}" -gt "${FHMAX_GFS}" ]] && export FHMAX_FITS=${FHMAX_GFS}
 
+# Fetch options
+export DO_FETCH_HPSS="@DO_FETCH_HPSS@"   # pull from HPSS
+export DO_FETCH_LOCAL="@DO_FETCH_LOCAL@"   # pull from local
```
These should probably just be `NO`.  If a user wants to fetch ICs, they will need to set this manually.  Maybe later, we can add an option to `setup_expt.py` that allows the user to specify a location/method to fetch ICs from, at which point templating these parameters will make sense.
```suggestion
export DO_FETCH_HPSS="NO"   # pull from HPSS
export DO_FETCH_LOCAL="NO"   # pull from local
```

DavidGrumm-NOAA replied:
Ok, changed to "NO"

DavidHuber-NOAA commented on parm/config/gfs/config.fetch:
```diff
@@ -0,0 +1,37 @@
+# Get task specific resources
+source "${EXPDIR}/config.resources" fetch
+
+export ICSDIR="@ICSDIR@" # User provided ICSDIR; blank if not provided
+export BASE_IC="@BASE_IC@" # Platform home for staged ICs
+
+export FETCH_YAML_TMPL="${PARMgfs}/stage/master_gfs.yaml.j2"
+export fetch_yaml="stage_atm_cold.yaml"
+
+# Set ICSDIR (if not defined)
+if [[ -z "${ICSDIR}" ]] ; then
+
+  ic_ver="20240610"
+
+  if (( NMEM_ENS > 0 )) ; then
+    ensic="${CASE_ENS}"
+  fi
+
+  if [[ "${DO_OCN:-NO}" == "YES" ]] ; then
+    ocnic="mx${OCNRES}"
+  fi
+
+  export ICSDIR="${BASE_IC}/${CASE}${ensic:-}${ocnic:-}/${ic_ver}"
+
+fi
+
+# Use of perturbations files for ensembles
+export USE_OCN_ENS_PERTURB_FILES="NO"
+export USE_ATM_ENS_PERTURB_FILES="NO"
```
All of the stage_ic variables should be removed.

```suggestion
export FETCH_YAML_TMPL="${PARMgfs}/fetch/C48_cold.yaml.j2"
```

DavidGrumm-NOAA replied:
Made those updates.

DavidHuber-NOAA commented on parm/fetch/stage_atm_cold.yaml:
```diff
@@ -0,0 +1,19 @@
+        {% endfor %} # ntile
+        {% endfor %} # ftype
+    destination: "{{ DATA }}"
+atmosphere_cold:
+    copy:
+        - ["{{ DATA }}/gfs_ctrl.nc", "{{ COMOUT_ATMOS_INPUT }}"]
+        {% for ftype in ["gfs_data", "sfc_data"] %}
+        {% for ntile in range(1, ntiles + 1) %}
+        - ["{{ DATA }}/{{ ftype }}.tile{{ ntile }}.nc", "{{ COMOUT_ATMOS_INPUT }}"]
+        {% endfor %} # ntile
+        {% endfor %} # ftype
```
We don't need to the copy portion.  This will be handled by the `stage_ic` job.

```suggestion
```

DavidHuber-NOAA replied:
Actually, this file should just be removed.  When `ci/cases/pr/ATM_cold.yaml` is moved here, it will replace this.

DavidGrumm-NOAA replied:
Removing parm/fetch/stage_atm_cold.yaml

DavidHuber-NOAA commented on workflow/applications/gfs_forecast_only.py:
```diff
@@ -27,9 +27,8 @@ def _get_app_configs(self, run):
         Returns the config_files that are involved in the forecast-only app
         """
-
+        configs = ['fetch', 'stage_ic', 'fcst', 'arch', 'cleanup']
         options = self.run_options[run]
```
`config.fetch` should only be added if `fetch_hpss` or `fetch_local` are `True`.

```suggestion
        configs = []
        options = self.run_options[run]
        if options['fetch_hpss'] or options['fetch_local']:
            configs += ['fetch']

        configs += ['stage_ic', 'fcst', 'arch', 'cleanup']
```

DavidGrumm-NOAA replied:
Made those modifications

DavidHuber-NOAA commented on ci/cases/pr/ATM_cold.yaml:
```diff
@@ -0,0 +1,11 @@
+untar:
+    tarball : "{{ ATARDIR }}/{{ cycle_YMDH }}/atm_cold.tar"
```
You may want to set up another directory on HPSS.  `ATARDIR` is where archived tarballs will be sent.  You could add a new directory to `config.base` and call it something different (`FETCHDIR`? or whatever fits your fancy).

DavidGrumm-NOAA replied:
Adding export for FETCHDIR in both versions of config.base

DavidHuber-NOAA commented on workflow/rocoto/gfs_tasks.py:
```diff
@@ -18,6 +18,13 @@ def _is_this_a_gdas_task(run, task_name):
     # Specific Tasks begin here
     def stage_ic(self):
 
+        deps = []
+        dep_dict = {
+            'type': 'task', 'name': f'fetch',
+        }
+        deps.append(rocoto.add_dependency(dep_dict))
+        dependencies = rocoto.create_dependency(dep=deps)
+
```
This should only be a dependency if the `fetch` job exists:

```suggestion
        if self.options['fetch_hpss'] or self.options['fetch_local']:
            deps = []
            dep_dict = {
                'type': 'task', 'name': f'fetch',
            }
            deps.append(rocoto.add_dependency(dep_dict))
            dependencies = rocoto.create_dependency(dep=deps)
```

DavidGrumm-NOAA replied:
Updated those lines.

DavidHuber-NOAA commented on workflow/rocoto/tasks.py:
```diff
@@ -10,7 +10,7 @@
 
 class Tasks:
-    SERVICE_TASKS = ['arch', 'earc', 'stage_ic', 'cleanup']
+    SERVICE_TASKS = ['arch', 'earc', 'fetch', 'cleanup']
```
`stage_ic` should stay here as well:

```suggestion
    SERVICE_TASKS = ['arch', 'earc', 'stage_ic', 'fetch', 'cleanup']
```

DavidGrumm-NOAA replied:
Made that change

DavidGrumm-NOAA committed:
Additonal code to incorporate fetch, including modifying some of the stage_ic and yaml code

DavidHuber-NOAA commented on parm/config/gefs/config.base:
```diff
@@ -103,6 +103,7 @@ export ROTDIR="@COMROOT@/${PSLOT}"
 
 export ARCDIR="${NOSCRUB}/archive/${PSLOT}"
 export ATARDIR="@ATARDIR@"
+export FETCHDIR="@FETCHDIR@"  # DG adding 121724
```
These comments are fine while testing, but they shouldn't be in the final PR.  Instead, a comment on what the `FETCHDIR` is for would be useful.

```suggestion
export FETCHDIR="@FETCHDIR@"  # HPSS or local directory where IC tarball(s) can be found.
```

DavidGrumm-NOAA replied:
I agree (done)

DavidHuber-NOAA commented on parm/config/gefs/config.fetch:
I think `parm/config/gefs/config.fetch` could be deleted and replaced with a link to `parm/config/gfs/config.fetch` since the contents should be the same.  That will save on maintaining the same script in two places.

DavidGrumm-NOAA replied:
Done - made it a hard link instead of a symlink (I thought this would be preferable in this case)

DavidHuber-NOAA commented on scripts/exglobal_fetch.py:
```diff
@@ -0,0 +1,45 @@
+    fetch = Fetch(config)
+
+    # Pull out all the configuration keys needed to run the fetch step
+    keys = ['current_cycle', 'RUN', 'PDY', 'PARMgfs', 'PSLOT', 'ROTDIR', 'fetch_yaml', 'ATARDIR', 'ntiles']
```
Now that you have a `FETCHDIR`, it should replace `ATARDIR` in the fetch job.
```suggestion
    keys = ['current_cycle', 'RUN', 'PDY', 'PARMgfs', 'PSLOT', 'ROTDIR', 'fetch_yaml', 'FETCHDIR', 'ntiles']
```

DavidGrumm-NOAA replied:
Right - done.

DavidHuber-NOAA commented on ush/python/pygfs/task/fetch.py:
```diff
@@ -0,0 +1,90 @@
+        return parsed_fetch
+
+    @logit(logger)
+    def execute_pull_data(self, atardir_set: Dict[str, Any]) -> None:
+        """Pull data from HPSS based on a yaml dictionary and store at the
+           specified destination.
+
+        Parameters
+        ----------
+        atardir_set: Dict[str, Any],
+            Dict defining set of tarballs to pull and where to put them.
+
+        Return
+            None
+        """
+
+        if len(f_names) <= 0:     # Abort if no files
+            raise FileNotFoundError("FATAL ERROR: The tar ball has no files")
+        f_names = atardir_set.untar.contents
+        on_hpss = atardir_set.untar.on_hpss
+        dest = atardir_set.untar.destination
```
Replace `atardir` with `fetchdir`:

```suggestion
    def execute_pull_data(self, fetchdir_set: Dict[str, Any]) -> None:
        """Pull data from HPSS based on a yaml dictionary and store at the
           specified destination.
        Parameters
        ----------
        fetchdir_set: Dict[str, Any],
            Dict defining set of tarballs to pull and where to put them.
        Return
            None
        """

        if len(f_names) <= 0:     # Abort if no files
            raise FileNotFoundError("FATAL ERROR: The tar ball has no files")
        f_names = fetchdir_set.untar.contents
        on_hpss = fetchdir_set.untar.on_hpss
        dest = fetchdir_set.untar.destination
```

DavidGrumm-NOAA replied:
Done

DavidHuber-NOAA commented on scripts/exglobal_fetch.py:
```diff
@@ -0,0 +1,45 @@
+
+    # Determine which archives to retrieve from HPSS
+    # Read the input YAML file to get the list of tarballs on tape
+    atardir_set = fetch.configure(fetch_dict)
+
+    # Pull the data from tape or locally and store the specified destination
+    fetch.execute_pull_data(atardir_set)
```
```suggestion
    fetchdir_set = fetch.configure(fetch_dict)

    # Pull the data from tape or locally and store the specified destination
    fetch.execute_pull_data(fetchdir_set)
```

DavidGrumm-NOAA replied:
Done

DavidHuber-NOAA commented on ci/cases/pr/ATM_cold.yaml:
```diff
@@ -0,0 +1,11 @@
+untar:
+    tarball : "{{ ATARDIR }}/{{ cycle_YMDH }}/atm_cold.tar"
```
This should be `FETCHDIR` now

```suggestion
    tarball : "{{ FETCHDIR }}/{{ cycle_YMDH }}/atm_cold.tar"
```

DavidGrumm-NOAA replied:
Done

DavidGrumm-NOAA committed:
Updated fetch code to use fetch directory FETCHDIR instead of ATARDIR, replaced gefs/config.fetch with a link to gfs/config.fetch

DavidGrumm-NOAA committed:
Deleted ci/cases/pr/ATM_cold.yaml

DavidGrumm-NOAA committed:
modifying stage_ic and fetch code to ensure that both function compatibly

DavidHuber-NOAA commented on scripts/exglobal_fetch.py:
```diff
@@ -0,0 +1,45 @@
+    fetch = Fetch(config)
+
+    # Pull out all the configuration keys needed to run the fetch step
+    keys = ['current_cycle', 'RUN', 'PDY', 'PARMgfs', 'PSLOT', 'ROTDIR', 'fetch_yaml', 'FETCHDIR', 'ntiles', 'DATAROOT', 'cycle_YMDH']
```
My bad, `cycle_YMDH` needs to be set in the jinja2 file.
```suggestion
    keys = ['current_cycle', 'RUN', 'PDY', 'PARMgfs', 'PSLOT', 'ROTDIR', 'fetch_yaml', 'FETCHDIR', 'ntiles', 'DATAROOT']
```

DavidHuber-NOAA commented on parm/fetch/ATM_cold.yaml.j2:
```diff
@@ -0,0 +1,11 @@
+untar:
```
```suggestion
{% set cycle_YMDH = current_cycle | to_YMDH %}
untar:
```

DavidGrumm-NOAA committed:
Moved setting of cycle_YMDH

DavidGrumm-NOAA commented:
Those changes fixed that ‘None’ error in the path, so it is now:

/NCEPDEV/emc-global/1year/David.Grumm/test_data/2021032312/atm_cold.tar

… but that directory ( /NCEPDEV/emc-global/1year/David.Grumm/test_data/2021032312/) does not seem to exist, so it still fails. From the log:

< snip >
024-12-23 20:30:21,057 - DEBUG    - fetch       :  returning: {'untar': {'tarball': '/NCEPDEV/emc-global/1year/David.Grumm/test_data/2021032312/atm_cold.tar', 'on_hpss': True, 'contents': ['gfs_ctrl.nc', 'gfs_data.tile1.nc', 'gfs_data.tile2.nc', 'gfs_data.tile3.nc', 'gfs_data.tile4.nc', 'gfs_data.tile5.nc', 'gfs_data.tile6.nc', 'sfc_data.tile1.nc', 'sfc_data.tile2.nc', 'sfc_data.tile3.nc', 'sfc_data.tile4.nc', 'sfc_data.tile5.nc', 'sfc_data.tile6.nc'], 'destination': '/scratch1/NCEPDEV/stmp2/David.Grumm/RUNDIRS/pslot_3141/gfs.2021032312'}}
2024-12-23 20:30:21,057 - INFO     - fetch       : BEGIN: pygfs.task.fetch.execute_pull_data
2024-12-23 20:30:21,057 - DEBUG    - fetch       : ( <pygfs.task.fetch.Fetch object at 0x148321730290>, {'untar': {'tarball': '/NCEPDEV/emc-global/1year/David.Grumm/test_data/2021032312/atm_cold.tar', 'on_hpss': True, 'contents': ['gfs_ctrl.nc', 'gfs_data.tile1.nc', 'gfs_data.tile2.nc', 'gfs_data.tile3.nc', 'gfs_data.tile4.nc', 'gfs_data.tile5.nc', 'gfs_data.tile6.nc', 'sfc_data.tile1.nc', 'sfc_data.tile2.nc', 'sfc_data.tile3.nc', 'sfc_data.tile4.nc', 'sfc_data.tile5.nc', 'sfc_data.tile6.nc'], 'destination': '/scratch1/NCEPDEV/stmp2/David.Grumm/RUNDIRS/pslot_3141/gfs.2021032312'}} )
[connecting to hpsscore1.fairmont.rdhpcs.noaa.gov/1217]
ERROR: [FATAL] no such HPSS archive file: /NCEPDEV/emc-global/1year/David.Grumm/test_data/2021032312/atm_cold.tar
###WARNING  htar returned non-zero exit status.
            72 = /apps/hpss_hera/bin/htar -x -v -f /NCEPDEV/emc-global/1year/David.Grumm/test_data/2021032312/atm_cold.tar gfs_ctrl.nc gfs_data.tile1.nc gfs_data.tile2.nc gfs_data.tile3.nc gfs_data.tile4.nc gfs_data.tile5.nc gfs_data.tile6.nc sfc_data.tile1.nc sfc_data.tile2.nc sfc_data.tile3.nc sfc_data.tile4.nc sfc_data.tile5.nc sfc_data.tile6.nc
HTAR: HTAR FAILED
WARNING: Unable to chdir(/scratch1/NCEPDEV/stmp2/David.Grumm/RUNDIRS/pslot_3141/gfs.2021032312)
Traceback (most recent call last):
  File "/scratch1/NCEPDEV/global/David.Grumm/G_WF_2988/scripts/exglobal_fetch.py", line 45, in <module>
    main()
  File "/scratch1/NCEPDEV/global/David.Grumm/G_WF_2988/ush/python/wxflow/logger.py", line 266, in wrapper
    retval = func(*args, **kwargs)
             ^^^^^^^^^^^^^^^^^^^^^
  File "/scratch1/NCEPDEV/global/David.Grumm/G_WF_2988/scripts/exglobal_fetch.py", line 41, in main
    fetch.execute_pull_data(fetchdir_set)
  File "/scratch1/NCEPDEV/global/David.Grumm/G_WF_2988/ush/python/wxflow/logger.py", line 266, in wrapper
    retval = func(*args, **kwargs)
             ^^^^^^^^^^^^^^^^^^^^^
  File "/scratch1/NCEPDEV/global/David.Grumm/G_WF_2988/ush/python/pygfs/task/fetch.py", line 89, in execute_pull_data
    htar_obj.xvf(tarball, f_names)
  File "/scratch1/NCEPDEV/global/David.Grumm/G_WF_2988/ush/python/wxflow/htar.py", line 174, in xvf
    output = self.extract(tarball, fileset, opts="-v")
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/scratch1/NCEPDEV/global/David.Grumm/G_WF_2988/ush/python/wxflow/htar.py", line 153, in extract
    output = self._htar(arg_list)
             ^^^^^^^^^^^^^^^^^^^^
  File "/scratch1/NCEPDEV/global/David.Grumm/G_WF_2988/ush/python/wxflow/htar.py", line 53, in _htar
    output = self.exe(*arg_list, output=str.split, error=str.split)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/scratch1/NCEPDEV/global/David.Grumm/G_WF_2988/ush/python/wxflow/executable.py", line 230, in __call__
    raise ProcessError(f"Command exited with status {proc.returncode}:", long_msg)
wxflow.executable.ProcessError: Command exited with status 72:
'/apps/hpss/htar' '-x' '-v' '-f' '/NCEPDEV/emc-global/1year/David.Grumm/test_data/2021032312/atm_cold.tar' 'gfs_ctrl.nc' 'gfs_data.tile1.nc' 'gfs_data.tile2.nc' 'gfs_data.tile3.nc' 'gfs_data.tile4.nc' 'gfs_data.tile5.nc' 'gfs_data.tile6.nc' 'sfc_data.tile1.nc' 'sfc_data.tile2.nc' 'sfc_data.tile3.nc' 'sfc_data.tile4.nc' 'sfc_data.tile5.nc' 'sfc_data.tile6.nc'
HTAR: HTAR FAILED
[connecting to hpsscore1.fairmont.rdhpcs.noaa.gov/1217]
ERROR: [FATAL] no such HPSS archive file: /NCEPDEV/emc-global/1year/David.Grumm/test_data/2021032312/atm_cold.tar
###WARNING  htar returned non-zero exit status.
            72 = /apps/hpss_hera/bin/htar -x -v -f /NCEPDEV/emc-global/1year/David.Grumm/test_data/2021032312/atm_cold.tar gfs_ctrl.nc gfs_data.tile1.nc gfs_data.tile2.nc gfs_data.tile3.nc gfs_data.tile4.nc gfs_data.tile5.nc gfs_data.tile6.nc sfc_data.tile1.nc sfc_data.tile2.nc sfc_data.tile3.nc sfc_data.tile4.nc sfc_data.tile5.nc sfc_data.tile6.nc

+ JGLOBAL_FETCH[1]: postamble JGLOBAL_FETCH 1734985804 1
+ preamble.sh[70]: set +x
End JGLOBAL_FETCH at 20:30:22 with error code 1 (time elapsed: 00:00:18)
+ fetch.sh[1]: postamble fetch.sh 1734985800 1
+ preamble.sh[70]: set +x
End fetch.sh at 20:30:22 with error code 1 (time elapsed: 00:00:22)
_______________________________________________________________
Start Epilog on node hfe01 for job 4228191 :: Mon Dec 23 20:30:23 UTC 2024
Job 4228191 finished for user David.Grumm in partition service with exit code 1:0
_______________________________________________________________
End Epilogue Mon Dec 23 20:30:23 UTC 2024

I will look into the intended creation of this directory.

DavidGrumm-NOAA commented:
The current file creation is correct - the error (“no such HPSS archive file”) is resolved by locating the tarball in the correct directory. There currently is a warning for the chdir call for htar/tar which I’m investigating.

DavidHuber-NOAA commented on scripts/exglobal_fetch.py:
```diff
@@ -0,0 +1,45 @@
+    fetch = Fetch(config)
+
+    # Pull out all the configuration keys needed to run the fetch step
+    keys = ['current_cycle', 'RUN', 'PDY', 'PARMgfs', 'PSLOT', 'ROTDIR', 'fetch_yaml', 'FETCHDIR', 'ntiles', 'DATAROOT']
```
`config.fetch` sets the name of the fetch job Jinja-YAML to `FETCH_YAML_TMPL`.  This is the environmental variable that `exglobal_fetch.py` should be looking for.

```suggestion
    keys = ['current_cycle', 'RUN', 'PDY', 'PARMgfs', 'PSLOT', 'ROTDIR', 'FETCH_YAML_TMPL', 'FETCHDIR', 'ntiles', 'DATAROOT']
```

DavidGrumm-NOAA replied:
Done

DavidHuber-NOAA commented on parm/config/gfs/config.fetch:
```diff
@@ -0,0 +1,12 @@
+
+echo "BEGIN: config.fetch"
+
+export FETCH_YAML_TMPL="${PARMgfs}/fetch/C48_cold.yaml.j2"
```
The name of the Jinja-YAML is `ATM_cold.yaml.j2`.

```suggestion
export FETCH_YAML_TMPL="${PARMgfs}/fetch/ATM_cold.yaml.j2"
```

DavidGrumm-NOAA replied:
Done

DavidHuber-NOAA commented on scripts/exglobal_fetch.py:
```diff
@@ -0,0 +1,45 @@
+        if fetch_dict[key] is None:
+            print(f"Warning: key ({key}) not found in task_config!")
+
+    # Also import all COMOUT* directory and template variables
+    for key in fetch.task_config.keys():
+        if key.startswith("COMOUT_"):
+            fetch_dict[key] = fetch.task_config.get(key)
+            if fetch_dict[key] is None:
+                print(f"Warning: key ({key}) not found in task_config!")
+
```
Thinking about it, the `fetch` task doesn't interact with the `COM` directories, so I think this can be deleted.

```suggestion
```

DavidGrumm-NOAA replied:
Done

DavidHuber-NOAA commented on workflow/rocoto/gfs_tasks.py:
```diff
@@ -18,6 +18,15 @@ def _is_this_a_gdas_task(run, task_name):
+        if self.options['do_fetch_hpss'] or self.options['do_fetch_local']:
+            deps = []
+            dep_dict = {
+                'type': 'task', 'name': f'fetch',
```
The task name should also have the `RUN` in it:

```suggestion
                'type': 'task', 'name': f'{self.run}_fetch',
```

DavidGrumm-NOAA replied:
Done

DavidHuber-NOAA commented on workflow/rocoto/tasks.py:
```diff
@@ -10,7 +10,7 @@
 class Tasks:
-    SERVICE_TASKS = ['arch', 'earc', 'stage_ic', 'cleanup']
+    SERVICE_TASKS = ['arch', 'earc', 'stage_ic', 'fetch', 'cleanup']
     VALID_TASKS = ['aerosol_init', 'stage_ic',
```
`fetch` should also be added to the `VALID_TASKS` list.

DavidGrumm-NOAA replied:
Done

DavidHuber-NOAA commented on ush/python/pygfs/task/fetch.py:
```diff
@@ -0,0 +1,95 @@
+        """
+        self.hsi = Hsi()
+
+        fetch_yaml = fetch_dict.fetch_yaml
```
This should be `FETCH_YAML_TMPL`:

```suggestion
        fetch_yaml = fetch_dict.FETCH_YAML_TMPL
```

DavidGrumm-NOAA replied:
Done

DavidGrumm-NOAA committed:
Updates for the fetch and yaml code, and keys.

DavidGrumm-NOAA committed:
Corrected typo

DavidGrumm-NOAA committed:
Add specification of fetch directory, corrrected name of yaml file.

DavidGrumm-NOAA committed:
Add DO_FETCH_HPSS setting to all non-hera files in workflow/hosts

DavidHuber-NOAA commented on ush/python/pygfs/task/fetch.py:
```diff
@@ -0,0 +1,94 @@
+                htar_obj = Htar.Htar()
+                htar_obj.xvf(tarball, f_names)
+            else:  # tar all files in fnames
+                pass  # TODO
```
Suggest replacing this with a `NotImplementedError`.

DavidGrumm-NOAA replied:
Done

DavidGrumm-NOAA committed:
Change pass to raise an error

DavidGrumm-NOAA committed:
whitespace repair

DavidGrumm-NOAA committed:
Merge branch 'develop' into stage_ic_2988

DavidHuber-NOAA committed:
Add S2SW_cold fetch template

DavidHuber-NOAA committed:
Update ATM_cold.yaml.j2

DavidHuber-NOAA committed:
Add YMDH to fetch_yamls

DavidHuber-NOAA committed:
Add previous_cycle to fetch dict

DavidHuber-NOAA committed:
Check for missing files after untarring, add logging

DavidHuber-NOAA committed:
Renamed fetch configs

DavidHuber-NOAA committed:
Replace gefs config.fetch with link to gfs

DavidHuber-NOAA committed:
Update fetch directory

DavidHuber-NOAA committed:
Restrict fetch cases to C48_S2SW and C48_ATM

DavidGrumm-NOAA committed:
Merge pull request #1 from DavidHuber-NOAA/new_fetch_yaml

Updates to fetch

DavidHuber-NOAA commented on workflow/hosts/googlepw.yaml:
```diff
@@ -24,7 +24,8 @@ LOCALARCH: 'NO'
-DO_GENESIS: 'NO'
-DO_METP: 'NO'
-SUPPORTED_RESOLUTIONS: ['C48', 'C96', 'C384'] # TODO: Test and support all cubed-sphere resolutions.
+DO_FETCH_HPSS: 'NO'
+SUPPORT_WAVES: 'NO'
+SUPPORTED_RESOLUTIONS: ['C48', 'C96'] # TODO: Test and support all cubed-sphere resolutions.
+
+
```
```suggestion
DO_TRACKER: 'NO'
DO_GENESIS: 'NO'
DO_METP: 'NO'
DO_FETCH_HPSS: 'NO'
SUPPORTED_RESOLUTIONS: ['C48', 'C96', 'C384'] # TODO: Test and support all cubed-sphere resolutions.
```

DavidGrumm-NOAA replied:
Done

DavidHuber-NOAA commented on workflow/hosts/azurepw.yaml:
```diff
@@ -24,7 +24,7 @@ LOCALARCH: 'NO'
-DO_GENESIS: 'NO'
-DO_METP: 'NO'
-SUPPORTED_RESOLUTIONS: ['C48', 'C96', 'C384', 'C768'] # TODO: Test and support all cubed-sphere resolutions.
+DO_FETCH_HPSS: 'NO' 
+SUPPORT_WAVES: 'NO'
+SUPPORTED_RESOLUTIONS: ['C48', 'C96'] # TODO: Test and support all cubed-sphere resolutions.
+
```
```suggestion
DO_TRACKER: 'NO'
DO_GENESIS: 'NO'
DO_METP: 'NO'
DO_FETCH_HPSS: 'NO' 
SUPPORTED_RESOLUTIONS: ['C48', 'C96', 'C384', 'C768'] # TODO: Test and support all cubed-sphere resolutions.
```

DavidGrumm-NOAA replied:
Done

DavidHuber-NOAA commented on workflow/hosts/awspw.yaml:
```diff
@@ -27,5 +27,7 @@ MAKE_ACFTBUFR: 'NO'
 DO_METP: 'NO'
-SUPPORTED_RESOLUTIONS: ['C48', 'C96', 'C192', 'C384', 'C768'] # TODO: Test and support all cubed-sphere resolutions.
-AERO_INPUTS_DIR: /contrib/global-workflow-shared-data/data/gocart_emissions
+DO_FETCH_HPSS: 'NO'
+SUPPORT_WAVES: 'NO'
+SUPPORTED_RESOLUTIONS: ['C48', 'C96'] # TODO: Test and support all cubed-sphere resolutions.
+
```
```suggestion
DO_FETCH_HPSS: 'NO'
SUPPORTED_RESOLUTIONS: ['C48', 'C96', 'C192', 'C384', 'C768'] # TODO: Test and support all cubed-sphere resolutions.
AERO_INPUTS_DIR: /contrib/global-workflow-shared-data/data/gocart_emissions
```

DavidGrumm-NOAA replied:
Done

DavidHuber-NOAA commented on ci/cases/yamls/gfs_defaults_ci.yaml:
```diff
@@ -1,5 +1,14 @@
   DO_TEST_MODE: "NO"
+  FETCHDIR: "/NCEPDEV/emc-global/1year/David.Huber/test_data"
+  DO_FETCH_HPSS: "YES"
+
```
```suggestion
```

DavidGrumm-NOAA replied:
Done

DavidHuber-NOAA reviewed (changes requested):
Some cleanup in the env files.

DavidGrumm-NOAA committed:
Cleanup of env files

DavidGrumm-NOAA committed:
Merge branch 'develop' into stage_ic_2988

DavidHuber-NOAA commented on ush/python/pygfs/task/fetch.py:
```diff
@@ -0,0 +1,105 @@
+        None
+        """
+        super().__init__(config)
+        # Perhaps add other stuff to self.
```
This was just a placeholder.  Doesn't look like anything was needed.
```suggestion
```

DavidGrumm-NOAA commented:
I have undrafted this PR so it is ready for review.

DavidHuber-NOAA commented on parm/fetch/gfs_ATM_cold_forecast-only.yaml.j2:
```diff
@@ -0,0 +1,16 @@
+{% set cycle_YMD = current_cycle | to_YMD %}
+{% set cycle_HH = current_cycle | strftime("%H") %}
+{% set atm_dir = RUN + "." ~ cycle_YMD ~ "/" ~ cycle_HH ~ "/model/atmos/input" %}
+untar:
```
Maybe instead of `untar`, this should be `target` or something else.  It is a little confusing in `fetch.py` when reading line 74:

```python
        f_names = fetchdir_set.untar.contents
```
Which makes it look like it is running an untar command.

DavidGrumm-NOAA replied:
So would the corresponding change be made in fetch.py, e,g,  : 

 f_names = fetchdir_set.target.contents
on_hpss = fetchdir_set.target.on_hpss
etc. ?

DavidHuber-NOAA replied:
Yes, that's correct.

DavidHuber-NOAA commented on ush/python/pygfs/task/fetch.py:
```diff
@@ -0,0 +1,105 @@
+                htar_obj = htar.Htar()
+                htar_obj.xvf(tarball, f_names)
+            else:  # tar all files in fnames
+                raise (NotImplementedError)
```
```suggestion
                raise NotImplementedError("The fetch job does not yet support pulling from local archives")
```

DavidHuber-NOAA commented on workflow/hosts/azurepw.yaml:
```diff
@@ -26,5 +26,8 @@ MAKE_NSSTBUFR: 'NO'
-DO_METP: 'NO'
+DO_METP: 'NO' 
+DO_FETCH_HPSS: 'NO' 
+SUPPORT_WAVES: 'NO'
 SUPPORTED_RESOLUTIONS: ['C48', 'C96', 'C384', 'C768'] # TODO: Test and support all cubed-sphere resolutions.
+
```
`SUPPORT_WAVES` isn't used anymore.

```suggestion
SUPPORTED_RESOLUTIONS: ['C48', 'C96', 'C384', 'C768'] # TODO: Test and support all cubed-sphere resolutions.
```

DavidHuber-NOAA commented on workflow/hosts/googlepw.yaml:
```diff
@@ -27,4 +27,8 @@ MAKE_ACFTBUFR: 'NO'
 DO_GENESIS: 'NO'
 DO_METP: 'NO'
+DO_FETCH_HPSS: 'NO'
+SUPPORT_WAVES: 'NO'
```
```suggestion
```

DavidHuber-NOAA reviewed (commented):
Looks good overall.  Just a few minor suggestions.

DavidGrumm-NOAA committed:
Renamed variable, removed unused code

DavidGrumm-NOAA committed:
need to pull changes from remote. Merge branch 'stage_ic_2988' of github.com:DavidGrumm-NOAA/Global_Workflow_2988 into stage_ic_2988

DavidHuber-NOAA reviewed (dismissed):
Looks good to me.

DavidHuber-NOAA commented:
@DavidGrumm-NOAA Apologies, I see now that there are a couple of conflicts.  Could you pull in develop and resolve the conflicts?  I think we can start CI testing after that.

DavidGrumm-NOAA committed:
Merged in develop

DavidGrumm-NOAA committed:
Merged in develop

DavidHuber-NOAA commented on workflow/hosts/gaeac5.yaml:
```diff
@@ -24,5 +24,7 @@ LOCALARCH: 'NO'
 MAKE_ACFTBUFR: 'NO'
-SUPPORTED_RESOLUTIONS: ['C384', 'C192', 'C96', 'C48']
+DO_FETCH_HPSS: 'NO'
+SUPPORTED_RESOLUTIONS: ['C1152', 'C768', 'C384', 'C192', 'C96', 'C48']
```
I think this got mangled in the merge.  This should be

```suggestion
SUPPORTED_RESOLUTIONS: ['C384', 'C192', 'C96', 'C48']
```

WalterKolczynski-NOAA replied:
I think the higher resolutions need to stay in the supported list.

DavidGrumm-NOAA replied:
Corrected that and repushed.

DavidGrumm-NOAA committed:
Undo merge mangling

DavidGrumm-NOAA commented:
Ok. I will revert those back. Also- Walter: For some reason I don't see
your comment on the PR page (only in the email)

On Thu, Jan 23, 2025 at 1:44 PM Walter Kolczynski - NOAA <
***@***.***> wrote:

> ***@***.**** commented on this pull request.
> ------------------------------
>
> In workflow/hosts/gaeac5.yaml
> <https://github.com/NOAA-EMC/global-workflow/pull/3141#discussion_r1927499034>
> :
>
> > @@ -24,5 +24,7 @@ LOCALARCH: 'NO'
>  ATARDIR: '${NOSCRUB}/archive_rotdir/${PSLOT}'
>  MAKE_NSSTBUFR: 'NO'
>  MAKE_ACFTBUFR: 'NO'
> -SUPPORTED_RESOLUTIONS: ['C384', 'C192', 'C96', 'C48']
> +DO_FETCH_HPSS: 'NO'
> +SUPPORTED_RESOLUTIONS: ['C1152', 'C768', 'C384', 'C192', 'C96', 'C48']
>
> I think the higher resolutions need to stay in the supported list.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/NOAA-EMC/global-workflow/pull/3141#discussion_r1927499034>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/BKF4TU5RNQ272366FH75ELD2ME2CHAVCNFSM6AAAAABTCZFW46VHI2DSMVQWIX3LMV43YUDVNRWFEZLROVSXG5CSMV3GSZLXHMZDKNZQG4ZDMNJYGU>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

WalterKolczynski-NOAA commented on ci/cases/yamls/gfs_defaults_ci.yaml:
```diff
@@ -1,6 +1,15 @@
 base:
   ACCOUNT: {{ 'HPC_ACCOUNT' | getenv }}
   DO_TEST_MODE: "NO"
+  FETCHDIR: "/NCEPDEV/emc-global/1year/David.Grumm/test_data"
+  DO_FETCH_HPSS: "YES"
```
I don't think we want to turn this on in CI tests.

DavidGrumm-NOAA replied:
Done

DavidGrumm-NOAA committed:
Adding higher resolutions back

WalterKolczynski-NOAA commented on parm/config/gfs/config.base:
```diff
@@ -474,6 +475,19 @@ export DO_VRFY_OCEANDA="@DO_VRFY_OCEANDA@"  # Run SOCA Ocean and Seaice DA verif
 [[ "${FHMAX_FITS}" -gt "${FHMAX_GFS}" ]] && export FHMAX_FITS=${FHMAX_GFS}
 
+# Fetch options
+# TODO once tarballs are available for CI at other resolutions/start types, etc, remove this if block
+if [[ "${APP}" == "ATM" || "${APP}" == "S2SW" ]] && \
+ [[ "${EXP_WARM_START}" == ".false." && "${CASE}" == "C48" ]]; then
+   export DO_FETCH_HPSS="@DO_FETCH_HPSS@"   # pull from HPSS  ## 123124 DG change to YES, or next
+   export DO_FETCH_LOCAL="NO"   # pull from local
+fi
```
I know this is temporary anyway, but I don't think it should be handled here at all.

DavidGrumm-NOAA replied:
Done

WalterKolczynski-NOAA commented on parm/config/gfs/config.fetch:
```diff
@@ -0,0 +1,19 @@
+
+export FETCH_YAML_TMPL="${PARMgfs}/fetch/${NET}_${APP}_${ic_type}_${MODE}.yaml.j2"
+
+# Get task specific resources
+source "${EXPDIR}/config.resources" fetch
```
To be consistent with other config files, sourcing resources should be first.

DavidGrumm-NOAA replied:
Done

WalterKolczynski-NOAA commented on workflow/rocoto/gfs_tasks.py:
```diff
@@ -29,6 +38,27 @@ def stage_ic(self):
+
+        return task
+
+    def fetch(self):
```
Move this task def above stage_ic so they appear in the logical order.

DavidGrumm-NOAA replied:
Done

DavidGrumm-NOAA committed:
Removing higher resolutions (they will be added in a different PR)

WalterKolczynski-NOAA commented on parm/fetch/gfs_S2SW_cold_forecast-only.yaml.j2:
```diff
@@ -0,0 +1,37 @@
+{% set cycle_YMDH = current_cycle | to_YMDH %}
+{% set cycle_YMD = current_cycle | to_YMD %}
+{% set cycle_HH = current_cycle | strftime("%H") %}
+{% set p_cycle_YMD = previous_cycle | to_YMD %}
```
Can we please use `previous_` or `prev_` instead of `p_` for these varnames?

DavidGrumm-NOAA replied:
Done, changed to "prev_"

DavidGrumm-NOAA replied:
I don't understand the RTD build failure "fatal: reference is not a tree: "

DavidHuber-NOAA replied:
It seems it couldn't find your commit hash.  I wouldn't worry about it yet.  It will probably resolve on the next commit.

DavidGrumm-NOAA committed:
Address additional reviewer comments

DavidHuber-NOAA commented on parm/config/gfs/config.base:
```diff
@@ -475,14 +475,6 @@ export DO_VRFY_OCEANDA="@DO_VRFY_OCEANDA@"  # Run SOCA Ocean and Seaice DA verif
-   export DO_FETCH_HPSS="@DO_FETCH_HPSS@"   # pull from HPSS  ## 123124 DG change to YES, or next
-   export DO_FETCH_LOCAL="NO"   # pull from local
-fi
-
```
I think the `DO_FETCH_HPSS` and `DO_FETCH_LOCAL` options should remain, but they should be set to `NO`.  It will be up to the user to enable these at experiment setup time.

```suggestion
export DO_FETCH_HPSS="NO"
export DO_FETCH_LOCAL="NO"
```

DavidGrumm-NOAA replied:
Done

DavidHuber-NOAA commented on ci/cases/yamls/gfs_defaults_ci.yaml:
```diff
@@ -10,6 +10,6 @@ base:
   DO_TEST_MODE: "NO"
   FETCHDIR: "/NCEPDEV/emc-global/1year/David.Grumm/test_data"
-  DO_FETCH_HPSS: "YES"
+  DO_FETCH_HPSS: "NO"
```
Since this will always be set to `NO`, there's no need to have it in the defaults.

DavidGrumm-NOAA replied:
Done

DavidHuber-NOAA commented on workflow/hosts/awspw.yaml:
```diff
@@ -27,5 +27,7 @@ MAKE_ACFTBUFR: 'NO'
 DO_TRACKER: 'NO'
 DO_GENESIS: 'NO'
 DO_METP: 'NO'
+DO_FETCH_HPSS: 'NO'
```
Similarly, for all host files, this entry is no longer required and should be removed.

DavidGrumm-NOAA replied:
Done for 11 yaml files

DavidGrumm-NOAA committed:
Remove some fetch options for now

DavidHuber-NOAA commented on ci/cases/yamls/gfs_defaults_ci.yaml:
```diff
@@ -1,6 +1,14 @@
+defaults:
+  !INC {{ HOMEgfs }}/parm/config/gfs/yaml/defaults.yaml
+base:
+  ACCOUNT: {{ 'HPC_ACCOUNT' | getenv }}
+  DO_TEST_MODE: "YES"
+
```
There are two sets of "defaults" and "base" in this yaml.  I think the first one should be removed.

```suggestion
```

DavidGrumm-NOAA replied:
Done

DavidHuber-NOAA replied:
~Lines 1-6 are still present.~

DavidHuber-NOAA replied:
Apologies, please disregard.

DavidHuber-NOAA commented on parm/config/gfs/config.base:
```diff
@@ -474,6 +475,10 @@ export DO_VRFY_OCEANDA="@DO_VRFY_OCEANDA@"  # Run SOCA Ocean and Seaice DA verif
 export FHMAX_FITS=132
 [[ "${FHMAX_FITS}" -gt "${FHMAX_GFS}" ]] && export FHMAX_FITS=${FHMAX_GFS}
 
+# User may choose to reset these at experiment setup time 
```
A comment on what these flags do is probably warranted.

DavidGrumm-NOAA replied:
Done

DavidGrumm-NOAA committed:
Address reviewer comments(.venv) [David.Grumm@hfe10 G_WF_2988]$ git add ci/cases/yamls/gfs_defaults_ci.yaml parm/config/gfs/config.base

DavidHuber-NOAA commented on workflow/hosts/awspw.yaml:
These host files have extra whitespace in them.  Could you please remove that?  There should be no difference between these and develop.

DavidGrumm-NOAA replied:
Are you referring to (for that file) a new line at the end ? I added that as I've read that UNIX tools expect a new line at the end. I can remove it from that file and others to agree with the develop versions.

DavidHuber-NOAA replied:
Yes, the new line character is expected, but most editors put this in automatically.  If, after removing these blank lines, we see artifacts indicating that there isn't a new line character at the end of the file, we can diagnose the issue in emacs.

DavidGrumm-NOAA replied:
Ok, thanks for the explanation. I will remove them.

DavidGrumm-NOAA replied:
Removed extraneous newlines.

DavidHuber-NOAA replied:
Try adding this to your .emacs file:
```.emacs
(setq require-final-newline 1)
```
Then open, edit (add, then delete a character), and save the file.  Lastly, run `xxd` on one of the files and verify that the last hex byte is `0A`.  This indicates that a new line was written at the end of the file.

DavidGrumm-NOAA committed:
Remove extraneous new lines

DavidHuber-NOAA committed:
Merge branch 'develop' into stage_ic_2988

DavidGrumm-NOAA committed:
Delete updated submodules to placate git pull. Merge branch 'stage_ic_2988' of github.com:DavidGrumm-NOAA/Global_Workflow_2988 into stage_ic_2988

DavidGrumm-NOAA commented:
I tried it - a nice, useful feature.  However, I may not always want there
to be a new line at the end. So although it seems that I can turn off the
prompt (when saving) asking if I want to append a newline, I may then
forget that that would be the default. I will read up on this some more.
Thanks.

On Fri, Jan 24, 2025 at 4:19 PM David Huber ***@***.***>
wrote:

> ***@***.**** commented on this pull request.
> ------------------------------
>
> On workflow/hosts/awspw.yaml
> <https://github.com/NOAA-EMC/global-workflow/pull/3141#discussion_r1929241469>
> :
>
> Try adding this to your .emacs file:
>
> (setq require-final-newline 1)
>
> Then open, edit (add, then delete a character), and save the file. Lastly,
> run xxd on one of the files and verify that the last hex byte is 0A. This
> indicates that a new line was written at the end of the file.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/NOAA-EMC/global-workflow/pull/3141#discussion_r1929241469>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/BKF4TU63STTAXYG5WL4WB332MKU7LAVCNFSM6AAAAABTCZFW46VHI2DSMVQWIX3LMV43YUDVNRWFEZLROVSXG5CSMV3GSZLXHMZDKNZTGU4TGMZZGY>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

DavidGrumm-NOAA commented:
Walter - After reconciling your comments with Dave Huber's, I think I have
made all of the agreed-upon changes from your review.

On Thu, Jan 23, 2025 at 2:28 PM Walter Kolczynski - NOAA <
***@***.***> wrote:

> ***@***.**** requested changes on this pull request.
> ------------------------------
>
> In ci/cases/yamls/gfs_defaults_ci.yaml
> <https://github.com/NOAA-EMC/global-workflow/pull/3141#discussion_r1927554886>
> :
>
> > +  FETCHDIR: "/NCEPDEV/emc-global/1year/David.Grumm/test_data"
> +  DO_FETCH_HPSS: "YES"
>
> I don't think we want to turn this on in CI tests.
> ------------------------------
>
> In parm/config/gfs/config.base
> <https://github.com/NOAA-EMC/global-workflow/pull/3141#discussion_r1927558169>
> :
>
> > +# TODO once tarballs are available for CI at other resolutions/start types, etc, remove this if block
> +if [[ "${APP}" == "ATM" || "${APP}" == "S2SW" ]] && \
> + [[ "${EXP_WARM_START}" == ".false." && "${CASE}" == "C48" ]]; then
> +   export ***@***.***_FETCH_HPSS@"   # pull from HPSS  ## 123124 DG change to YES, or next
> +   export DO_FETCH_LOCAL="NO"   # pull from local
> +fi
>
> I know this is temporary anyway, but I don't think it should be handled
> here at all.
> ------------------------------
>
> In parm/config/gfs/config.fetch
> <https://github.com/NOAA-EMC/global-workflow/pull/3141#discussion_r1927561482>
> :
>
> > +# Get task specific resources
> +source "${EXPDIR}/config.resources" fetch
>
> To be consistent with other config files, sourcing resources should be
> first.
> ------------------------------
>
> In workflow/rocoto/gfs_tasks.py
> <https://github.com/NOAA-EMC/global-workflow/pull/3141#discussion_r1927565753>
> :
>
> > @@ -29,6 +38,27 @@ def stage_ic(self):
>                       'command': f'{self.HOMEgfs}/jobs/rocoto/stage_ic.sh',
>                       'job_name': ***@***.***',
>                       'log': ***@***.***@***@***.***@H/{task_name}.log',
> +                     'maxtries': '&MAXTRIES;',
> +                     'dependency': dependencies
> +                     }
> +
> +        task = rocoto.create_task(task_dict)
> +
> +        return task
> +
> +    def fetch(self):
>
> Move this task def above stage_ic so they appear in the logical order.
> ------------------------------
>
> In parm/fetch/gfs_S2SW_cold_forecast-only.yaml.j2
> <https://github.com/NOAA-EMC/global-workflow/pull/3141#discussion_r1927574178>
> :
>
> > @@ -0,0 +1,37 @@
> +{% set cycle_YMDH = current_cycle | to_YMDH %}
> +{% set cycle_YMD = current_cycle | to_YMD %}
> +{% set cycle_HH = current_cycle | strftime("%H") %}
> +{% set p_cycle_YMD = previous_cycle | to_YMD %}
>
> Can we please use 'previous_orprev_instead ofp_` for these varnames?
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/NOAA-EMC/global-workflow/pull/3141#pullrequestreview-2570802497>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/BKF4TUZE62VLZ3P5AWYVTKD2ME7FDAVCNFSM6AAAAABTCZFW46VHI2DSMVQWIX3LMV43YUDVNRWFEZLROVSXG5CSMV3GSZLXHMZDKNZQHAYDENBZG4>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

DavidGrumm-NOAA committed:
Add newline to end of yaml files

DavidHuber-NOAA reviewed (approved):
Changes look good.  Thanks @DavidGrumm-NOAA!!

DavidHuber-NOAA committed:
Merge branch 'develop' into stage_ic_2988
